# Lab 3 - Part 2: PCA and Clustering (12 marks)
### Due Date: Monday, March 13 at 12pm

Author: Jashraj Dubal

UCID: 300478888

The purpose of this portion of the assignment is to practice using PCA and clustering techniques on a given dataset

In [1]:
import numpy as np
import pandas as pd

## 0. Function definitions (2 marks)

In [3]:
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score

def cluster_fn(n_clusters, X, n_components=0):
    '''Calculate silhouette score for a given dataset, number of clusters, 
       and number of principle components using Kmeans clustering (random_state=0)
        
        n_clusters (int): number of clusters to use for Kmeans
        n_components (int): number of principle components (optional)
        X (numpy.array or pandas.DataFrame): unlabelled dataset
        
        returns: silhouette score
    
    '''
    # TODO: Implement function body
    # Get the silhouette score for the KMeans clustering
    if(n_components > 0):
        X = PCA(n_components=n_components).fit_transform(X)
    
    kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(X)
    return silhouette_score(X, kmeans.labels_)

## 1. Load data (2 marks)

For this assignment, we will use the dataset found below:

https://archive.ics.uci.edu/ml/datasets/Chemical+Composition+of+Ceramic+Samples

In [6]:
# TODO: Import dataset
data = pd.read_csv('Chemical Composion of Ceramic.csv')

Two of the columns are non-numeric. For this assignment, we will remove those two columns and focus on clustering the ceramic samples based on the numerical measurements

In [7]:
data

Ceramic Name   Part  Na2O   MgO  Al2O3   SiO2   K2O    CaO  TiO2  Fe2O3  \
0       FLQ-1-b   Body  0.62  0.38  19.61  71.99  4.84   0.31  0.07   1.18   
1       FLQ-2-b   Body  0.57  0.47  21.19  70.09  4.98   0.49  0.09   1.12   
2       FLQ-3-b   Body  0.49  0.19  18.60  74.70  3.47   0.43  0.06   1.07   
3       FLQ-4-b   Body  0.89  0.30  18.01  74.19  4.01   0.27  0.09   1.23   
4       FLQ-5-b   Body  0.03  0.36  18.41  73.99  4.33   0.65  0.05   1.19   
..          ...    ...   ...   ...    ...    ...   ...    ...   ...    ...   
83     DY-M-3-g  Glaze  0.34  0.55  12.37  70.70  5.33   8.06  0.06   1.61   
84    DY-QC-1-g  Glaze  0.72  0.34  12.20  72.19  6.19   6.06  0.04   1.27   
85    DY-QC-2-g  Glaze  0.23  0.24  12.99  71.81  5.25   7.15  0.05   1.29   
86    DY-QC-3-g  Glaze  0.14  0.46  12.62  69.16  4.34  11.03  0.05   1.20   
87    DY-QC-4-g  Glaze  0.14  0.63  14.25  71.55  4.87   6.43  0.08   1.05   

     MnO  CuO  ZnO  PbO2  Rb2O  SrO  Y2O3  ZrO2  P2O5  
0    630   10   70    10   430    0    40    80    90  
1    380   20   80    40   430  -10    40   100   110  
2    420   20   50    50   380   40    40    80   200  
3    460   20   70    60   380   10    40    70   210  
4    380   40   90    40   360   10    30    80   150  
..   ...  ...  ...   ...   ...  ...   ...   ...   ...  
83  1250   10   90    30   250  520    30   140   690  
84  1700   60  110    10   270  540    40   120   630  
85   750   40  100     0   240  470    40   120   480  
86   920   40   90    20   230  470    40   130  1100  
87   800   40   90    20   220  410    40   110   690  

[88 rows x 19 columns]

In [8]:
# TODO: Remove non-numeric columns
data.drop(['Ceramic Name', 'Part'], axis=1, inplace=True)

In [9]:
data

Na2O   MgO  Al2O3   SiO2   K2O    CaO  TiO2  Fe2O3   MnO  CuO  ZnO  PbO2  \
0   0.62  0.38  19.61  71.99  4.84   0.31  0.07   1.18   630   10   70    10   
1   0.57  0.47  21.19  70.09  4.98   0.49  0.09   1.12   380   20   80    40   
2   0.49  0.19  18.60  74.70  3.47   0.43  0.06   1.07   420   20   50    50   
3   0.89  0.30  18.01  74.19  4.01   0.27  0.09   1.23   460   20   70    60   
4   0.03  0.36  18.41  73.99  4.33   0.65  0.05   1.19   380   40   90    40   
..   ...   ...    ...    ...   ...    ...   ...    ...   ...  ...  ...   ...   
83  0.34  0.55  12.37  70.70  5.33   8.06  0.06   1.61  1250   10   90    30   
84  0.72  0.34  12.20  72.19  6.19   6.06  0.04   1.27  1700   60  110    10   
85  0.23  0.24  12.99  71.81  5.25   7.15  0.05   1.29   750   40  100     0   
86  0.14  0.46  12.62  69.16  4.34  11.03  0.05   1.20   920   40   90    20   
87  0.14  0.63  14.25  71.55  4.87   6.43  0.08   1.05   800   40   90    20   

    Rb2O  SrO  Y2O3  ZrO2  P2O5  
0    430    0    40    80    90  
1    430  -10    40   100   110  
2    380   40    40    80   200  
3    380   10    40    70   210  
4    360   10    30    80   150  
..   ...  ...   ...   ...   ...  
83   250  520    30   140   690  
84   270  540    40   120   630  
85   240  470    40   120   480  
86   230  470    40   130  1100  
87   220  410    40   110   690  

[88 rows x 17 columns]

## 2. Implement clustering (8 marks)

### 2.1 Cluster using raw data (1 mark)

Implement Kmeans clustering using the raw data. Compare the silhouette scores using 2, 3, 4, 5 and 6 clusters

In [23]:
# TODO: Implement clustering with raw data using cluster_fn above
clustering = pd.DataFrame()
clusters = [2,3,4,5,6]
for cluster in clusters:
    clustering = pd.concat([clustering, pd.DataFrame({'Clusters': cluster, 'Silhouette Score': cluster_fn(cluster, data)}, index=[0])], ignore_index=True)
    print("Silhouette score for {} clusters is {}".format(cluster, cluster_fn(cluster, data)))

Silhouette score for 2 clusters is 0.5840130686182087
Silhouette score for 3 clusters is 0.5616399840165864
Silhouette score for 4 clusters is 0.543410860697891
Silhouette score for 5 clusters is 0.5080642704990367
Silhouette score for 6 clusters is 0.5103991098981858


### 2.2 Cluster using PCA-transformed data (2 marks)

Implement Kmeans clustering using the PCA-transformed data. Compare the silhouette scores using 2, 3, 4, 5 and 6 clusters and 2, 3, 4, 5 and 6 principle components 

In [24]:
# TODO: Implement clustering with PCA-transformed data using cluster_fn above
pca_clustering = pd.DataFrame()
clusters = [2,3,4,5,6]
principle_components = [2,3,4,5,6]
for cluster in clusters:
    for component in principle_components:
        # Add the silhouette score to the dataframe with its corresponding cluster and component
        pca_clustering = pd.concat([pca_clustering, pd.DataFrame({'Cluster': cluster, 'Component': component, 'Silhouette Score': cluster_fn(cluster, data, component)}, index=[0])], ignore_index=True)
        print("Silhouette score for {} clusters and {} components is {}".format(cluster, component, cluster_fn(cluster, data, component)))

Silhouette score for 2 clusters and 2 components is 0.6194422739820585
Silhouette score for 2 clusters and 3 components is 0.5999612865384474
Silhouette score for 2 clusters and 4 components is 0.5899549664230885
Silhouette score for 2 clusters and 5 components is 0.5874723776389513
Silhouette score for 2 clusters and 6 components is 0.585962969580958
Silhouette score for 3 clusters and 2 components is 0.6116248458476962
Silhouette score for 3 clusters and 3 components is 0.5866089718929353
Silhouette score for 3 clusters and 4 components is 0.5709488853404948
Silhouette score for 3 clusters and 5 components is 0.5674695588445687
Silhouette score for 3 clusters and 6 components is 0.5647248342959946
Silhouette score for 4 clusters and 2 components is 0.6007517657297899
Silhouette score for 4 clusters and 3 components is 0.5705309604281563
Silhouette score for 4 clusters and 4 components is 0.5537151283872483
Silhouette score for 4 clusters and 5 components is 0.5492856440931448
Silhoue

### 2.3 Display results (2 marks)

Print the results for 2.1 and 2.2 in a table. Include column and row labels

In [25]:
# TODO: Display results
display(clustering)
display(pca_clustering)

Clusters  Silhouette Score
0         2          0.584013
1         3          0.561640
2         4          0.543411
3         5          0.508064
4         6          0.510399

Cluster  Component  Silhouette Score
0         2          2          0.619442
1         2          3          0.599961
2         2          4          0.589955
3         2          5          0.587472
4         2          6          0.585963
5         3          2          0.611625
6         3          3          0.586609
7         3          4          0.570949
8         3          5          0.567470
9         3          6          0.564725
10        4          2          0.600752
11        4          3          0.570531
12        4          4          0.553715
13        4          5          0.549286
14        4          6          0.546752
15        5          2          0.567088
16        5          3          0.545911
17        5          4          0.521348
18        5          5          0.515809
19        5          6          0.512537
20        6          2          0.569320
21        6          3          0.550951
22        6          4          0.529829
23        6          5          0.524216
24        6          6          0.515274

**Question**: Which combination of number of clusters and number of components produced the best results? What is the silhouette score for this combination? **(3 marks)**

The ideal silhoutte score is 1, so the closest combination is 2 clusters and 2 components. This silhouette score for this combination is 0.619442.

## 3. Improve results (Bonus - 3 marks)

Think about how you could improve the results from the previous section. Two potential methods include preprocessing the data or selecting a different clustering algorithm. Repeat section 2 with your selected improvement method to determine what the new silhouette scores would be

In [ ]:
# TODO: Repeat steps 2.1-2.3 using a different method/preprocessing/etc.

**Question**: Why did you select this improvement method? Which combination of number of clusters and number of components produced the best results? Did you improve the silhouette scores? If yes, how much of an improvement did you get over the previous results?

*ANSWER HERE*